File: frontend/src/types/analytics.ts
export interface AnalyticsOverview {
  total_input_tokens: number;
  total_output_tokens: number;
  total_cache_creation_tokens: number;
  total_cache_read_tokens: number;
  total_sessions: number;
  total_tasks_with_sessions: number;
  estimated_cost_usd: number;
  active_sessions_today: number;
}

export interface DailyTokens {
  date: string;
  input_tokens: number;
  output_tokens: number;
  cache_creation_tokens: number;
  cache_read_tokens: number;
}

export interface WeeklyTokens {
  week_start: string;
  input_tokens: number;
  output_tokens: number;
  cache_creation_tokens: number;
  cache_read_tokens: number;
}

export interface MonthlyTokens {
  month: string;
  input_tokens: number;
  output_tokens: number;
  cache_creation_tokens: number;
  cache_read_tokens: number;
}

export interface TaskTokens {
  task_id: string;
  task_title: string;
  input_tokens: number;
  output_tokens: number;
  cache_creation_tokens: number;
  cache_read_tokens: number;
  total_tokens: number;
}

export interface SessionTokens {
  session_id: string;
  task_title: string;
  input_tokens: number;
  output_tokens: number;
  cache_creation_tokens: number;
  cache_read_tokens: number;
  total_tokens: number;
  started_at: string | null;
}

export interface ToolTokens {
  tool_name: string;
  input_tokens: number;
  output_tokens: number;
  call_count: number;
}

export interface LanguageTokens {
  file_ext: string;
  input_tokens: number;
  output_tokens: number;
  call_count: number;
}

export interface EfficiencyRow {
  task_id: string;
  task_title: string;
  total_tokens: number;
  lines_written: number;
  project_loc: number;
  tokens_per_line: number | null;
  tokens_per_loc: number | null;
}

export interface SessionTimelineEvent {
  sequence_no: number;
  event_type: string;
  tool_name: string | null;
  input_tokens: number;
  output_tokens: number;
  cumulative_total: number;
  timestamp: string;
}

export interface TaskTimelineEvent {
  claude_session_id: string;
  sequence_no: number;
  event_type: string;
  tool_name: string | null;
  input_tokens: number;
  output_tokens: number;
  cumulative_total: number;
  timestamp: string;
}

export interface UsageWindows {
  tokens_5hr: number;
  tokens_week: number;
  limit_5hr: number;
  limit_week: number;
  reset_5hr: string | null;
  reset_week: string | null;
  no_data: boolean;
}

export interface CostByTask {
  task_id: string;
  task_title: string;
  input_tokens: number;
  output_tokens: number;
  cache_creation_tokens: number;
  cache_read_tokens: number;
  cost_usd: number;
}

export interface TokensByStage {
  stage: string;
  input_tokens: number;
  output_tokens: number;
}

export interface SessionSummary {
  total_sessions: number;
  avg_tokens_per_session: number;
  max_tokens_per_session: number;
  total_cost_usd: number;
}

export interface BurnRate {
  tokens_last_hour: number;
  tokens_per_minute: number;
}

export interface SessionDetail {
  id: string;
  task_id: string;
  status: string;
  started_at: string;
  ended_at: string | null;
  claude_session_id: string | null;
  error_message: string | null;
  duration_secs: number | null;
  input_tokens: number;
  output_tokens: number;
  total_tokens: number;
}

export interface DevActivityRow {
  task_id: string;
  task_title: string;
  session_count: number;
  lines_added: number;
  lines_deleted: number;
  input_tokens: number;
  output_tokens: number;
  cache_creation_tokens: number;
  cache_read_tokens: number;
  cost_usd: number;
}

export interface PlanTier {
  tier: string;
  limit_5hr: number;
  limit_week: number;
}

export interface RoiMetrics {
  cost_per_commit: number | null;
  cost_per_pr: number | null;
  cost_per_loc: number | null;
  total_commits: number;
  total_prs: number;
  total_loc: number;
  total_active_time_secs: number;
  avg_session_duration_secs: number;
  total_cost_usd: number;
}

export interface ContextWindowUsage {
  session_id: string;
  task_title: string;
  tokens_in_window: number;
  context_limit: number;
  pct_used: number;
}

export interface LocHistoryEntry {
  session_id: string;
  session_index: number;
  project_loc: number;
  started_at: string;
}

export interface HeatmapEntry {
  date: string;   // "YYYY-MM-DD"
  tokens: number;
}

export interface HourlyEntry {
  hour: number;   // 0–23
  tokens: number;
}

// Session-scoped tool breakdown uses the same shape as ToolTokens
export type SessionToolTokens = ToolTokens;

File: backend/src/models/analytics.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize)]
pub struct AnalyticsOverview {
    pub total_input_tokens: i64,
    pub total_output_tokens: i64,
    pub total_sessions: i64,
    pub total_tasks_with_sessions: i64,
    pub estimated_cost_usd: f64,
    pub active_sessions_today: i64,
}

#[derive(Debug, Serialize)]
pub struct DailyTokens {
    pub date: String,
    pub input_tokens: i64,
    pub output_tokens: i64,
}

#[derive(Debug, Serialize)]
pub struct WeeklyTokens {
    pub week_start: String,
    pub input_tokens: i64,
    pub output_tokens: i64,
}

#[derive(Debug, Serialize)]
pub struct MonthlyTokens {
    pub month: String,
    pub input_tokens: i64,
    pub output_tokens: i64,
}

#[derive(Debug, Serialize)]
pub struct TaskTokens {
    pub task_id: String,
    pub task_title: String,
    pub input_tokens: i64,
    pub output_tokens: i64,
    pub total_tokens: i64,
}

#[derive(Debug, Serialize)]
pub struct SessionTokens {
    pub session_id: String,
    pub task_title: String,
    pub input_tokens: i64,
    pub output_tokens: i64,
    pub total_tokens: i64,
    pub started_at: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct ToolTokens {
    pub tool_name: String,
    pub input_tokens: i64,
    pub output_tokens: i64,
    pub call_count: i64,
}

#[derive(Debug, Serialize)]
pub struct LanguageTokens {
    pub file_ext: String,
    pub input_tokens: i64,
    pub output_tokens: i64,
    pub call_count: i64,
}

#[derive(Debug, Serialize)]
pub struct EfficiencyRow {
    pub task_id: String,
    pub task_title: String,
    pub total_tokens: i64,
    pub lines_written: i64,
    pub project_loc: i64,
    pub tokens_per_line: Option<f64>,
    pub tokens_per_loc: Option<f64>,
}

/// Token usage summary for rate-limit windows (5-hour and weekly).
/// `limit_*` fields are 0 when not configured (no limit known).
#[derive(Debug, Serialize, Deserialize)]
pub struct UsageWindows {
    pub tokens_5hr: i64,
    pub tokens_week: i64,
    pub limit_5hr: i64,
    pub limit_week: i64,
    /// ISO-8601 timestamp when the current 5-hr window resets (null if no usage)
    pub reset_5hr: Option<String>,
    /// ISO-8601 timestamp when the weekly window resets (always next Monday 00:00 UTC)
    pub reset_week: String,
}

#[derive(Debug, Serialize)]
pub struct SessionTimelineEvent {
    pub sequence_no: i64,
    pub event_type: String,
    pub tool_name: Option<String>,
    pub input_tokens: i64,
    pub output_tokens: i64,
    pub cumulative_total: i64,
    pub timestamp: String,
}

#[derive(Debug, Serialize)]
pub struct TaskTimelineEvent {
    pub claude_session_id: String,
    pub sequence_no: i64,
    pub event_type: String,
    pub tool_name: Option<String>,
    pub input_tokens: i64,
    pub output_tokens: i64,
    pub cumulative_total: i64,
    pub timestamp: String,
}

#[derive(Debug, Serialize)]
pub struct CostByTask {
    pub task_id: String,
    pub task_title: String,
    pub input_tokens: i64,
    pub output_tokens: i64,
    pub cost_usd: f64,
}

#[derive(Debug, Serialize)]
pub struct TokensByStage {
    pub stage: String,
    pub input_tokens: i64,
    pub output_tokens: i64,
}

#[derive(Debug, Serialize)]
pub struct SessionSummary {
    pub total_sessions: i64,
    pub avg_tokens_per_session: f64,
    pub max_tokens_per_session: i64,
    pub total_cost_usd: f64,
}

#[derive(Debug, Serialize)]
pub struct BurnRate {
    pub tokens_last_hour: f64,
    pub tokens_per_minute: f64,
}

#[derive(Debug, Serialize)]
pub struct SessionDetail {
    pub id: String,
    pub task_id: String,
    pub status: String,
    pub started_at: String,
    pub ended_at: Option<String>,
    pub claude_session_id: Option<String>,
    pub error_message: Option<String>,
    pub duration_secs: Option<i64>,
    pub input_tokens: i64,
    pub output_tokens: i64,
    pub total_tokens: i64,
}

#[derive(Debug, Serialize)]
pub struct PlanTier {
    pub tier: String,
    pub limit_5hr: i64,
    pub limit_week: i64,
}

#[derive(Debug, Serialize)]
pub struct RoiMetrics {
    pub cost_per_commit:   Option<f64>,
    pub cost_per_pr:       Option<f64>,
    pub cost_per_loc:      Option<f64>,
    pub total_commits:     i64,
    pub total_prs:         i64,
    pub total_loc:         i64,
    pub total_active_time_secs: f64,
    pub avg_session_duration_secs: f64,
    pub total_cost_usd:    f64,
}
